TemporaryProjects/StartPageExtender.cs

```diff
@@ -0,0 +1,158 @@
+                    windowPane?.GetType().GetProperty(nameof(WindowPane.Content)).GetValue(windowPane) is IVsUIWpfElement wpfElement &&
+                    ErrorHandler.Succeeded(wpfElement.GetFrameworkElement(out var element)) &&
+                    element is FrameworkElement frameworkElement &&
+                    frameworkElement.FindName("ContentHost") is Decorator contentHost)
+                {
+                    SetupStartPage(contentHost);
+                }
+            }
+        }
+
+        private void SetupStartPage(Decorator contentHost)
+        {
+            contentHost.LayoutUpdated += LayoutUpdated;
+
+            void LayoutUpdated(object sender, EventArgs e)
+            {
+                var recentProjectsPanel = FindRecentProjectsPanel(contentHost);
+                if (recentProjectsPanel != null)
+                {
+                    AddNewTemporaryProjectButton(recentProjectsPanel);
+                    contentHost.LayoutUpdated -= LayoutUpdated;
+                }
+            }
+        }
+
+        private static Grid FindRecentProjectsPanel(Decorator contentHost)
+        {
+            if (contentHost.Child is ContentControl start &&
+                start.Template.FindName("NewProjectsListView", start) is ContentControl newProjectsListView &&
+                newProjectsListView.Template.FindName("RecentProjectsPanel", newProjectsListView) is Grid recentProjectsPanel)
+            {
+                return recentProjectsPanel;
+            }
+
+            return null;
+        }
+
+        private void AddNewTemporaryProjectButton(Grid recentProjectsPanel)
+        {
+            var button = new NewTemporaryProjectButton(dte)
+            {
+                Style = ((Button)recentProjectsPanel.FindName("MoreTemplatesButton")).Style
+            };
+
+            recentProjectsPanel.RowDefinitions.Add(new RowDefinition() { Height = GridLength.Auto });
+            Grid.SetRow(button, recentProjectsPanel.RowDefinitions.Count - 1);
+
+            recentProjectsPanel.Children.Add(button);
+        }
+
+        private sealed class NewTemporaryProjectButton : Button
+        {
+            private readonly DTE dte;
+
+            public NewTemporaryProjectButton(DTE dte)
+            {
+                this.dte = dte;
+            }
+
+            public override void OnApplyTemplate()
+            {
+                base.OnApplyTemplate();
+
+                var textBlock = (TextBlock)GetTemplateChild("MoreTemplatesText");
+                textBlock.Text = "Create new temporary project...";
+            }
+
+            protected override void OnClick()
+            {
+                ThreadHelper.ThrowIfNotOnUIThread();
+
+                base.OnClick();
+
+                object _ = null;
+                dte.Commands.Raise(NewTempProjectCommand.CommandSet.ToString("B"), NewTempProjectCommand.CommandId, _, _);
```

Comment:
🤔 Huh, the API says it takes 2 ref parameters but it compiles without the `ref` keyword. Not only that, but I was able to pass in a `null` literal directly. I guess I still have a lot to learn about C# (or COM?).

Reply:
They seem to behave like `in` parameters, even though the IDE shows `ref`.

Reply:
This is a COM hack that I learnt about recently. Apparently there are a few COM APIs that use `ref` parameters pretty much everywhere, but only use them as input parameters (they don't mutate them). This makes using the API bad enough that they introduced this hack to allow `ref` parameters to be treated like input parameters for COM objects.

I didn't know you could pass in a `null` literal directly!  I'd be inclined to do that.

Reply:
> Apparently there are a few COM APIs that use ref parameters pretty much everywhere, but only use them as input parameters

Hm, that seems weird given that `object` is a reference type and passing it by ref shouldn't really by more efficient. But I don't understand how this works under the hood.